### PR TITLE
Remove unused attention_mask from entire call stack

### PR DESCRIPTION
**Pull request by haok1402**
1. flash-attn-4 for GQA in Qwen3
2. TileLang/Triton for MLA in DeepSeek-V2-Lite

Neither of the attention kernels in this codebase require the "materialized" attention mask. However, we do create it and pass down the entire call stack, which is unnecessary. Training does gets launched properly after this removal.

```
2026-03-29 20:02:21 | INFO | launch(cfg=PretrainLanguageModelCfg(distributed=DistributedCfg(pipeline_parallel_size=2, context_parallel_size=1, expert_parallel_size=2), training=TrainingCfg(dataset=PosixPath('workspace/pithtrain/datasets/dclm-baseline/toktxt/deepseek-v2'), sequence_length=4096, seed=1234, min_lr=1e-06, max_lr=1e-06, warmup_steps=0, max_steps=25, micro_batch_size=1, global_batch_size=1024, optimizer='Adam', scheduler='Constant', model=PosixPath('Pith-Train/examples/pretrain_language_model/deepseek-v2-lite/config.json'), save_interval=25, save_location=PosixPath('workspace/pithtrain/checkpoints/deepseek-v2-lite'), moe_load_balance_coef=0.003, moe_load_balance_type='sequence', fp8_training='disabled', init_std=0.02, nsys_start=None, nsys_stop=None), logging=LoggingCfg(wandb=None)))
2026-03-29 20:02:31 | INFO | Load checkpoint: workspace/pithtrain/checkpoints/deepseek-v2-lite/torch-dcp/step-00000000
2026-03-29 20:02:37 | INFO | Load checkpoint: Elapsed min=5.9s, max=5.9s
2026-03-29 20:05:34 | INFO | step 00000001/00000025 | step-time 177.036 sec | cross-entropy-loss 2.5317 | load-balance-loss 0.003228 | learning-rate 1.000000e-06 | gradient-norm 6.1188 | tokens-per-second 23,692 | peak-gpu-memory 95.47 GB
2026-03-29 20:06:17 | INFO | step 00000002/00000025 | step-time 43.126 sec | cross-entropy-loss 2.5194 | load-balance-loss 0.003223 | learning-rate 1.000000e-06 | gradient-norm 6.1281 | tokens-per-second 97,256 | peak-gpu-memory 95.43 GB
2026-03-29 20:07:01 | INFO | step 00000003/00000025 | step-time 43.226 sec | cross-entropy-loss 2.5041 | load-balance-loss 0.003225 | learning-rate 1.000000e-06 | gradient-norm 5.9667 | tokens-per-second 97,032 | peak-gpu-memory 95.50 GB
```

### CODE REVIEW

File: tests/test_fsdp.py [hunk 1]
```diff
@@ -279,7 +268,7 @@ def causal(b, h, q_idx, kv_idx):
     kwargs = dict()
     kwargs["pp_group"] = pp_group
     kwargs["ep_group"] = ep_group
-    kwargs["const_inputs"] = (attention_mask,)
+    kwargs["const_inputs"] = ()
```
[the remark this answers is not in the document]
Comment (haok1402) [reply tests/test_fsdp.py:271]: @gemini-code-assist

**Before this PR**, `const_inputs=(attention_mask,)`. In `overlapped_forward_backward`:

```python
attention_mask = const_inputs0[0]
position_ids = const_inputs0[1] if len(const_inputs0) > 1 else None  # → None
```

`position_ids` was **already `None`** in the overlapped pipeline path. Our PR didn't change that. We only removed `attention_mask` from index 0 — `position_ids` was never in `const_inputs` to begin with.

The two code paths handle `position_ids=None` correctly:

1. **Non-overlapped path** (`_forward_compute_chunk` → `model.forward()`): The model's `forward()` creates `position_ids` with the correct CP offset internally before passing it to each layer.

2. **Overlapped path** (`overlapped_forward_backward` → `stage1_f` → `layer.forward_attn`): `position_ids=None` is passed to the attention module. DeepSeek handles this with `cos[None]` broadcasting (positions 0..S-1). Qwen3 uses `layer._position_embeddings` set during the model-level forward, and doesn't consume `position_ids` in `_forward_attn_compute` at all.

This was the **identical behavior before our PR**. Gemini incorrectly attributed a pre-existing design choice as a regression introduced by this change.

File: tests/test_deepseek_backward_compat.py [hunk 2]
```diff
@@ -105,14 +104,7 @@ def main():
...
 
-        def causal(b, h, q_idx, kv_idx):
-            return q_idx >= kv_idx
-
-        attention_mask = create_block_mask(causal, B=None, H=None, Q_LEN=S, KV_LEN=S)
-
-        model = DualPipeV(
-            modules, const_inputs=(attention_mask,), pp_group=pp_group, ep_group=ep_group
-        )
+        model = DualPipeV(modules, const_inputs=(), pp_group=pp_group, ep_group=ep_group)
```
[the remark this answers is not in the document]
Comment (haok1402) [reply tests/test_deepseek_backward_compat.py:107]: @gemini-code-assist

**Before this PR**, `const_inputs=(attention_mask,)`. In `overlapped_forward_backward`:

```python
attention_mask = const_inputs0[0]
position_ids = const_inputs0[1] if len(const_inputs0) > 1 else None  # → None
```

`position_ids` was **already `None`** in the overlapped pipeline path. Our PR didn't change that. We only removed `attention_mask` from index 0 — `position_ids` was never in `const_inputs` to begin with.

The two code paths handle `position_ids=None` correctly:

1. **Non-overlapped path** (`_forward_compute_chunk` → `model.forward()`): The model's `forward()` creates `position_ids` with the correct CP offset internally before passing it to each layer.

2. **Overlapped path** (`overlapped_forward_backward` → `stage1_f` → `layer.forward_attn`): `position_ids=None` is passed to the attention module. DeepSeek handles this with `cos[None]` broadcasting (positions 0..S-1). Qwen3 uses `layer._position_embeddings` set during the model-level forward, and doesn't consume `position_ids` in `_forward_attn_compute` at all.

This was the **identical behavior before our PR**. Gemini incorrectly attributed a pre-existing design choice as a regression introduced by this change.

File: pithtrain/modules/training.py [hunk 3]
```diff
@@ -360,9 +354,7 @@ def causal(b, h, q_idx, kv_idx):
                     if cp_group is not None:
                         gate.compute = gate.compute.__wrapped__.__get__(gate, type(gate))
 
-    ctx.model = DualPipeV(
-        modules, const_inputs=(attention_mask,), pp_group=pp_group, ep_group=ep_group
-    )
+    ctx.model = DualPipeV(modules, const_inputs=(), pp_group=pp_group, ep_group=ep_group)
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/modules/training.py:357]: @gemini-code-assist

**Before this PR**, `const_inputs=(attention_mask,)`. In `overlapped_forward_backward`:

```python
attention_mask = const_inputs0[0]
position_ids = const_inputs0[1] if len(const_inputs0) > 1 else None  # → None
```

`position_ids` was **already `None`** in the overlapped pipeline path. Our PR didn't change that. We only removed `attention_mask` from index 0 — `position_ids` was never in `const_inputs` to begin with.

The two code paths handle `position_ids=None` correctly:

1. **Non-overlapped path** (`_forward_compute_chunk` → `model.forward()`): The model's `forward()` creates `position_ids` with the correct CP offset internally before passing it to each layer.

2. **Overlapped path** (`overlapped_forward_backward` → `stage1_f` → `layer.forward_attn`): `position_ids=None` is passed to the attention module. DeepSeek handles this with `cos[None]` broadcasting (positions 0..S-1). Qwen3 uses `layer._position_embeddings` set during the model-level forward, and doesn't consume `position_ids` in `_forward_attn_compute` at all.

This was the **identical behavior before our PR**. Gemini incorrectly attributed a pre-existing design choice as a regression introduced by this change.

**Review by MasterJH5574 — APPROVED** [review]
Thanks for removing the legacy
